File: src/main/java/me/pcasaes/hexoids/domain/utils/TrigUtil.java
package me.pcasaes.hexoids.domain.utils;

public class TrigUtil {

    public static final float PI = (float) Math.PI;

    public static final float HALF_CIRCLE_IN_RADIANS = (float) Math.PI;

    public static final float FULL_CIRCLE_IN_RADIANS = 2f * (float) Math.PI;


    private TrigUtil() {
    }

    /**
     * https://stackoverflow.com/a/30887154
     *
     * @param a
     * @param b
     * @return
     */
    public static float calculateAngleDistance(float a, float b) {
        float abDiff = a - b;

        float d = Math.abs(abDiff) % FULL_CIRCLE_IN_RADIANS;
        float r = d > HALF_CIRCLE_IN_RADIANS ? FULL_CIRCLE_IN_RADIANS - d : d;


        return (abDiff >= 0 && abDiff <= HALF_CIRCLE_IN_RADIANS) ||
                (abDiff <= -HALF_CIRCLE_IN_RADIANS && abDiff >= -FULL_CIRCLE_IN_RADIANS) ? r : -r;
    }

    public static float limitRotation(float currentAngle, float nextAngle, float maxAngleDelta) {

        /*
        Optimization of TrigUtil#calculateAngleDistance

        if the abs of the angle difference is is less that max angle delta than it's good enough to
        answer with nextAngle
         */
        float abDiff = nextAngle - currentAngle;
        float abDiffAbs = Math.abs(abDiff);
        if (abDiffAbs <= maxAngleDelta) {
            return nextAngle;
        }

        float d = abDiffAbs % FULL_CIRCLE_IN_RADIANS;
        float r = d > HALF_CIRCLE_IN_RADIANS ? FULL_CIRCLE_IN_RADIANS - d : d;


        float aDiff1 = (abDiff >= 0 && abDiff <= HALF_CIRCLE_IN_RADIANS) ||
                (abDiff <= -HALF_CIRCLE_IN_RADIANS && abDiff >= -FULL_CIRCLE_IN_RADIANS) ? r : -r;


        if (aDiff1 > maxAngleDelta) {
            aDiff1 = maxAngleDelta;
            return currentAngle + aDiff1;
        } else if (aDiff1 < -maxAngleDelta) {
            aDiff1 = -maxAngleDelta;
            return currentAngle + aDiff1;
        } else {
            return nextAngle;
        }
    }

    /**
     * https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line
     * <p>
     * (x1, y1) -> (x2, y2) => two point in a line
     * (x, y) => point from which to get distance to the line
     *
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     * @param x
     * @param y
     * @return distance
     */
    public static float calculateShortestDistanceFromPointToLine(float x1, float y1,
                                                                 float x2, float y2,
                                                                 float x, float y) {
        return Math.abs((y2 - y1) * x - (x2 - x1) * y + x2 * y1 - y2 * x1) /
                (float) Math.sqrt(Math.pow(y2 - y1, 2.) + Math.pow(x2 - x1, 2.));
    }

    /**
     * Calculates the angle in radians between two points.
     *
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     * @return
     */
    public static float calculateAngleBetweenTwoPoints(float x1, float y1,
                                                       float x2, float y2) {
        return (float) Math.atan2(y2 - y1, x2 - x1);
    }

    public static float calculateAngleFromComponents(float x, float y) {
        return (float) Math.atan2(y, x);
    }

    public static float calculateMagnitudeFromComponents(float x, float y) {
        return (float) Math.sqrt(x * x + y * y);
    }

    public static float calculateXComponentFromAngleAndMagnitude(float angle, float speed) {
        return speed * (float) Math.cos(angle);
    }

    public static float calculateYComponentFromAngleAndMagnitude(float angle, float speed) {
        return speed * (float) Math.sin(angle);
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/core/domain/utils/TrigUtil.kt
package me.pcasaes.hexoids.core.domain.utils

import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

object TrigUtil {
    const val PI: Float = Math.PI.toFloat()

    const val HALF_CIRCLE: Float = PI

    const val QUARTER_CIRCLE: Float = PI / 2F

    const val FULL_CIRCLE: Float = 2F * PI


    /**
     * https://stackoverflow.com/a/30887154
     *
     * @param a
     * @param b
     * @return
     */
    fun angleDistance(a: Float, b: Float): Float {
        val delta = (a - b + PI) % FULL_CIRCLE - PI
        return if (delta < -PI) delta + FULL_CIRCLE else delta
    }

    fun limitRotation(currentAngle: Float, nextAngle: Float, maxAngleDelta: Float): Float {
        /*
               Optimization of TrigUtil#calculateAngleDistance
       
               if the abs of the angle difference is is less that max angle delta than it's good enough to
               answer with nextAngle
                */

        val abDiff = nextAngle - currentAngle
        val abDiffAbs = abs(abDiff)
        if (abDiffAbs <= maxAngleDelta) {
            return nextAngle
        }

        val delta = angleDistance(nextAngle, currentAngle)
        val clampedDelta = delta.coerceIn(-maxAngleDelta, maxAngleDelta)
        return currentAngle + clampedDelta

    }

    /**
     * https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line
     *
     *
     * (x1, y1) -> (x2, y2) => two point in a line
     * (x, y) => point from which to get distance to the line
     *
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     * @param x
     * @param y
     * @return distance
     */
    fun shortestDistanceFromPointToLine(
        x1: Float, y1: Float,
        x2: Float, y2: Float,
        x: Float, y: Float
    ): Float {
        // divide by zero guard
        return if (x1.isSame(x2) && y1.isSame(y2)) {
            Float.POSITIVE_INFINITY
        } else {
            abs((y2 - y1) * x - (x2 - x1) * y + x2 * y1 - y2 * x1) / sqrt(
                (y2 - y1).toDouble().pow(2.0) + (x2 - x1).toDouble().pow(2.0)
            ).toFloat()
        }
    }

    /**
     * Calculates the angle in radians between two points.
     *
     * @param x1
     * @param y1
     * @param x2
     * @param y2
     * @return
     */
    fun angleBetweenTwoPoints(
        x1: Float, y1: Float,
        x2: Float, y2: Float
    ): Float {
        return atan2((y2 - y1).toDouble(), (x2 - x1).toDouble()).toFloat()
    }

    fun angleFromComponents(x: Float, y: Float): Float {
        return atan2(y.toDouble(), x.toDouble()).toFloat()
    }

    fun magnitudeFromComponents(x: Float, y: Float): Float {
        return sqrt((x * x + y * y).toDouble()).toFloat()
    }

    fun xComponentFromAngleAndMagnitude(angle: Float, speed: Float): Float {
        return speed * cos(angle.toDouble()).toFloat()
    }

    fun yComponentFromAngleAndMagnitude(angle: Float, speed: Float): Float {
        return speed * sin(angle.toDouble()).toFloat()
    }
}
